# Workflow details for the Culex biting miRNA analysis

## Upstream 
Information about sequence reads (raw read count, read alignment rate, etc.) for the data can be found [here]()

### Data Accession
Data was generated by collaborator...

The raw reads are available in NCBIâ€™s short read archive (SRA) under accession number XXXXYYYY

### Preprocessing and Quality Control

Trimmomatic (version 0.39) was used to trim sequence reads based on quality ([script](https://github.com/srmarzec/Culex_Biting_RNAseq/blob/main/Upstream/trim.sh))

FastQC (v0.11.9) was used for quality control visualization ([script](https://github.com/srmarzec/Culex_Biting_RNAseq/blob/main/Upstream/fastqc.sh))


Preliminary trimming and fastqc showed a poor "per sequence base content" for the first ~15 bases. We decided to use HEADCROP flag to remove the first 15 bases. All other flags (TRAILING, SLIDINGWINDOW, and MINLEN) are rather general/default for basic quality of bases and did not result in much difference of trimming.

Most samples have warnings or fail for "per sequence GC content" and "sequence duplication levels", which I expect for RNAseq data. An interesting occurence, using the HEADCROP flag to remove the first 15 bases results in all samples failing "per sequence tile quality", while trimming without HEADCROP results only in a warning for this. This is likely because the samples are now being measured over 135b instead of 150b, which I guess makes this check fail for sequence tile quality.

*not sure about this ask in meeting, because the script I have in my notes have different settings? Can be found in my notes: https://github.com/AngelaZhou779/RISE/blob/main/Notes.md and also here: https://github.com/AngelaZhou779/RISE/blob/main/script/Trimmomatic.sh


#### Cleaning out other small RNAs

Angela add here
Refer to https://github.com/AngelaZhou779/RISE/blob/main/Notes.md

#### Size sorting

We wanted to only keep reads that fell within 18 - 24 bases which is what we consider the size of miRNAs. This was done with a custom [python script](https://github.com/srmarzec/Culex_Biting_miRNA/blob/main/scripts/python_scripts/trimANDsizeSort.py) for all the files with this [script](https://github.com/srmarzec/Culex_Biting_miRNA/blob/main/scripts/sortSize_multi.sh)

### miRDeep2 
#### Index with Bowtie

Mapping was done using the *Culex quinquefasciatus* reference genome ([GCF_015732765.1](https://www.ncbi.nlm.nih.gov/assembly/GCF_015732765.1/)) found on NCBI

The genome was indexed for miRDeep2 using Bowtie (v1.3.1) with this [script](https://github.com/srmarzec/Culex_Biting_miRNA/blob/main/scripts/genome_index.sh). Note, for miRDeep2 to run properly, there can be no whitespace in the headers of the reference genome fasta so I had to remove/replace the whitespaces before indexing.

#### Run miRDeep2 to count reads for each miRNA

miRDeep2 (miRDeep2.0.1.3) was installed and run in a conda virtual environment which was created following this [markdown](https://github.com/srmarzec/Culex_Biting_miRNA/blob/main/misc/Conda_VirtualEnvironment.md).

miRDeep2 was [run](https://github.com/srmarzec/Culex_Biting_miRNA/blob/main/scripts/miRDeep2.sh) and we found no novel miRNAs. We had a more extensive starting list because of a recent publication. 

We then mapped reads to the known miRNAs with this [script](https://github.com/srmarzec/Culex_Biting_miRNA/blob/main/scripts/miRDeep_mapper.sh). This produced a count matrix we used for downstream analysis. 


## Downstream

All downstream analysis done in R using R version **4.0.2**

### DESeq
Using DESeq2 (v1.30.1) ([script](https://github.com/srmarzec/Culex_Biting_RNAseq/blob/main/Downstream/DESeq.R))
or 
https://github.com/AngelaZhou779/RISE/blob/main/script/DESeqmiRNA.R

### Target Predicition 


No significant GO term enrichment or KEGG pathway enrichment based on the few consenesus genes that were found between both target prediction software. 

### Obtaining CpipJ IDs from past assembly for gene IDs of the newest assembly
Because we are working with the newest Culex quinquefasciatus assembly from NCBI (GCF_015732765.1), we ran into issues where downstream analysis for GO and KEGG pathway enrichment were hindered since those databases had annotations based on past Culex quinquefasciatus assemblies using locus tags (i.e. the CpipJ IDs as identifiers). Details on how we obtained locus tags is [here](https://github.com/srmarzec/Culex_Biting_RNAseq/blob/main/misc/GeneID_LocusTag_Conversion.md).
